tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,753 @@
+				}
+			}
+			chkCRC32.Reset()
+			startIdx := idx
+			seriesRef := binary.BigEndian.Uint64(bs.Range(idx, idx+8))
+			idx += 8
+
+			mint := int64(binary.BigEndian.Uint64(bs.Range(idx, idx+8)))
+			idx += 8
+
+			maxt := int64(binary.BigEndian.Uint64(bs.Range(idx, idx+8)))
+			idx += 8
+
+			chunkRef := uint64(seq)<<32 | uint64(idx)
+			if err := f(seriesRef, chunkRef, mint, maxt); err != nil {
+				return &CorruptionErr{
+					Dir:       cdm.dir.Name(),
+					FileIndex: seq,
+					Err:       err,
+				}
+			}
+
+			idx++ // Skip encoding.
+			// Skip the data.
+			dataLen, n := binary.Uvarint(bs.Range(idx, idx+MaxChunkLengthFieldSize))
+			idx += n + int(dataLen)
+
+			// Check CRC.
+			sum := bs.Range(idx, idx+CRCSize)
+			if _, err := chkCRC32.Write(bs.Range(startIdx, idx)); err != nil {
+				return err
+			}
+			if act := chkCRC32.Sum(nil); !bytes.Equal(act, sum) {
+				return errors.Errorf("checksum mismatch expected:%x, actual:%x", sum, act)
+			}
+
+			idx += CRCSize
+		}
+		if idx > sliceLen {
+			// It should be equal to the slice length.
+			return &CorruptionErr{
+				Dir:       cdm.dir.Name(),
+				FileIndex: seq,
+				Err:       errors.Errorf("head chunk file doesn't include enough bytes to read the last chunk data - required:%v, available:%v, file:%d", idx, sliceLen, seq),
+			}
+		}
+	}
+
+	return nil
+}
+
+// Truncate deletes the head chunk files which are strictly below the mint.
+// mint should be in milliseconds.
+func (cdm *ChunkDiskMapper) Truncate(mint int64) error {
+	var removedFiles []int
+
+	cdm.mmappedChunksMtx.RLock()
+	for seq, bs := range cdm.mmappedChunkFiles {
+		if seq == cdm.curFileSequence {
+			continue
+		}
+		b := bs.Range(HeaderMaxtOffset, HeaderMaxtOffset+8)
+		maxt := binary.BigEndian.Uint64(b)
+		if int64(maxt) < mint {
+			removedFiles = append(removedFiles, seq)
+		}
+	}
+	cdm.mmappedChunksMtx.RUnlock()
+
+	return cdm.deleteFiles(removedFiles)
+}
+
+func (cdm *ChunkDiskMapper) deleteFiles(removedFiles []int) error {
+	closers := make([]io.Closer, 0, len(removedFiles))
+	cdm.mmappedChunksMtx.Lock()
+	for _, seq := range removedFiles {
+		cdm.size -= int64(cdm.mmappedChunkFiles[seq].Len())
+		closers = append(closers, cdm.closers[seq])
+		delete(cdm.mmappedChunkFiles, seq)
+		delete(cdm.closers, seq)
+		if err := os.Remove(segmentFile(cdm.dir.Name(), seq)); err != nil {
+			cdm.mmappedChunksMtx.Unlock()
+			var merr tsdb_errors.MultiError
+			merr.Add(err)
+			merr.Add(closeAll(closers))
+			return merr.Err()
+		}
+	}
+	cdm.mmappedChunksMtx.Unlock()
+
+	return closeAll(closers)
+}
+
+// Repair deletes all the head chunk files after the one which had the corruption
+// (including the corrupt file).
+func (cdm *ChunkDiskMapper) Repair(originalErr error) error {
+	err := errors.Cause(originalErr) // So that we can pick up errors even if wrapped.
+	cerr, ok := err.(*CorruptionErr)
```

Comment:
Do we need expose publically `CorruptionErr`?

Reply:
Maybe not. I will make it internal for now.